File: 01.Pattern-Sliding-Window/01.introduction/sliding-window/AverageOfSubarrayOfSizeK.java
import java.util.Arrays;

public class AverageOfSubarrayOfSizeK {
    public static double[] findAverages(int K, int[] arr) {
        double[] result = new double[arr.length - K + 1];
        double windowSum = 0;
        double windowStart = 0;
        for (int windowEnd = 0; windowEnd < arr.length; windowEnd++) {
            windowSum += arr[windowEnd];    // add the next element
            // slide the window, we don't need to slide if we've not hit the required window size of K
            if (windowEnd >= K - 1) {
                result[windowStart] = windowSum / K;    // calculate the average
                windowSum -= arr[windowStart];  // substract the element going out
                windowStart++;  // slice the window ahead
            }
        }


        return result
    }

    public static void main(String[] args) {
        double[] result = AverageOfSubarrayOfSizeK.findAverages(5, new int[] { 1, 3, 2, 6, -1, 4, 1, 8, 2 });

        System.out.prinltn("Average of subarrays of size K:" + Arrays.toString(result));
    }

}

File: 01.Pattern-Sliding-Window/01.introduction/brute-force/AverageOfSubarrayOfSizeK.cpp
using namespace std;

#include <iostream>
#include <vector>

class AverageOfSubarrayOfSizeK
{
public:
    static vector<double> findAverages(int K, const vector<int> &arr)
    {
        vector<double> result(arr.size() - K + 1);
        for (int i = 0; i <= arr.size() - K; i++)
        {
            // find sum of next 'K' elements
            double sum = 0;
            for (int j = i; j < i + K; j++)
            {
                sum += arr[j];
            }
            result[i] = sum / K; // calculate average
        }

        return result;
    }
};

// g++ -std=c++11 AverageOfSubarrayOfSizeK.cpp -o AverageOfSubarrayOfSizeK.out
int main(int argc, char *argv[])
{
    // const std::vector<int> ints = {1, 3, 2, 6, -1, 4, 1, 8, 2};
    // vector<double> result = AverageOfSubarrayOfSizeK::findAverages(5, ints);
    vector<double> result = AverageOfSubarrayOfSizeK::findAverages(5, vector<int>{1, 3, 2, 6, -1, 4, 1, 8, 2});
    cout << "Average of subarrays of size K: ";
    for (double d : result)
    {
        cout << d << " ";
    }

    cout << endl;
}